src/syntax/ast.rs

```diff
@@ -1589,29 +1595,110 @@ impl AstNode for Param {
+pub enum PatternKind {
+    /// A single identifier: `x`.
+    Ident(Ident),
+    /// A destructuring pattern: `(x, _, ..y)`.
+    Destructure(Vec<DestructuringKind>),
+}
+
+impl Pattern {
+    /// TODO (Marmare): comment and maybe rename
+    pub fn bindings(&self) -> PatternKind {
+        if self.0.children().len() > 1 {
+            let mut bindings = Vec::new();
+            for child in self.0.children() {
+                match child.kind() {
+                    SyntaxKind::Ident => {
+                        bindings.push(DestructuringKind::Ident(child.cast().unwrap()))
+                    }
+                    SyntaxKind::Spread => bindings
+                        .push(DestructuringKind::Sink(child.cast_first_match().unwrap())),
+                    _ => {}
+                }
+            }
+            PatternKind::Destructure(bindings)
+        } else if let Some(ident) = self.0.cast_first_match::<Ident>() {
+            PatternKind::Ident(ident)
+        } else {
+            PatternKind::Ident(Ident::default())
+        }
+    }
+
+    // Returns a list of all identifiers in the pattern.
+    pub fn idents(&self) -> Vec<Ident> {
+        match self.bindings() {
+            PatternKind::Ident(ident) => vec![ident],
+            PatternKind::Destructure(bindings) => bindings
+                .into_iter()
+                .filter_map(|binding| match binding {
+                    DestructuringKind::Ident(ident) => Some(ident),
+                    DestructuringKind::Sink(ident) => Some(ident),
+                })
+                .collect(),
+        }
+    }
+}
+
 node! {
     /// A let binding: `let x = 1`.
     LetBinding
 }
 
+pub enum LetBindingKind {
+    /// A normal binding: `let x = 1`.
+    Normal(Pattern),
+    /// A closure binding: `let f(x) = 1`.
+    Closure(Ident),
+}
+
+impl LetBindingKind {
+    // Returns a list of all identifiers in the pattern.
+    pub fn idents(&self) -> Vec<Ident> {
+        match self {
+            LetBindingKind::Normal(pattern) => pattern.idents(),
+            LetBindingKind::Closure(ident) => {
+                vec![ident.clone()]
+            }
+        }
+    }
+}
+
 impl LetBinding {
     /// The binding to assign to.
-    pub fn binding(&self) -> Ident {
-        match self.0.cast_first_match() {
-            Some(Expr::Ident(binding)) => binding,
-            Some(Expr::Closure(closure)) => closure.name().unwrap_or_default(),
-            _ => Ident::default(),
+    pub fn binding(&self) -> LetBindingKind {
+        if let Some(pattern) = self.0.cast_first_match::<Pattern>() {
+            LetBindingKind::Normal(pattern)
+        } else if let Some(closure) = self.0.cast_first_match::<Closure>() {
+            if let Some(ident) = closure.name() {
+                LetBindingKind::Closure(ident)
+            } else {
+                LetBindingKind::Closure(Ident::default())
+            }
+        } else {
+            LetBindingKind::Normal(Pattern::default())
         }
```

Comment:
can this be written simpler like above? if not, oh well, but it feels like it can.

Reply:
Should be a lot cleaner now.